pytorch_lightning/utilities/data.py

```diff
@@ -18,28 +18,55 @@
 from torch.utils.data import DataLoader, IterableDataset
 
 from pytorch_lightning.utilities import rank_zero_warn
+from pytorch_lightning.utilities.warnings import WarningCache
 
 BType = Union[torch.Tensor, str, Mapping[Any, "BType"], Iterable["BType"]]
 
+warning_cache = WarningCache()
+
+
+def _extract_batch_size(batch: BType, batch_sizes) -> bool:
```

Comment:
isn't there a more elegant implementation using a Generator (yield)? 
Then there is no special return value needed, and the warning can be handled externally.

Reply:
I'll try